module/apmfasthttp/closer.go

```diff
@@ -0,0 +1,37 @@
+package apmfasthttp
+
+import (
+	"github.com/valyala/fasthttp"
+	"go.elastic.co/apm"
+)
+
+// NewTxCloser returns a transaction closer.
+func NewTxCloser(ctx *fasthttp.RequestCtx, tx *apm.Transaction, bc *apm.BodyCapturer) *TxCloser {
```

Comment:
Why is it needed for Atreugo? How TxCloser will be used?

Is it just to access the `*apm.Transaction` and `*apm.BodyCapturer`, or do you still need to be able to end the transaction explicitly? If it's just to access the transaction: can `apm.TransactionFromContext` be used instead?

Reply:
```go
func (m *Middleware) Middleware() atreugo.Middleware {
	return func(ctx *atreugo.RequestCtx) error {
		if !m.tracer.Recording() || m.requestIgnorer(ctx) {
			return ctx.Next()
		}

		tx, bc, err := apmfasthttp.StartTransactionWithBody(m.tracer, m.requestName(ctx), ctx.RequestCtx)
		if err != nil {
			return err
		}

		ctx.SetUserValue(txKey, apmfasthttp.NewTxCloser(ctx.RequestCtx, tx, bc))

		return ctx.Next() // nolint:wrapcheck
	}
}
```

Reply:
I will put the `*apm.Transaction` into the user values like `apmfasthttp`, that will end the transacction explicitly.

Reply:
I see, thanks. A few thoughts:

- it's not very obvious (to me at least) that you would need to call `ctx.SetUserValue(txKey, apmfasthttp.NewTxCloser(ctx.RequestCtx, tx, bc))` before `apm.TransactionFromContext(ctx.RequestCtx)` will work
- should `apmfasthttp.StartTransactionWithBody` call `SetUserValue`? I think it probably should
- could we alternatively hide the implementation behind a function which returns `io.Closer`?

My gut feeling is that the simplest thing to do would be to have StartTransactionWithBody call SetUserValue itself, rather than forcing the caller to do that. Then the TxCloser can be unexported, and consumers don't need to know about it.

Reply:
Yes, i can move the `ctx.SetUserValue` into `apmfasthttp.StartTransactionWithBody` but equally i need the `TxCloser` to be public due to in the recovery function i need:

```go
func (m *Middleware) PanicView(t *apm.Tracer) atreugo.PanicView {
	return func(ctx *atreugo.RequestCtx, err interface{}) {
		if m.panicPropagation {
			defer panic(err)
		}

		// 500 status code will be set only for APM transaction
		// to allow other middleware to choose a different response code
		if ctx.Response.Header.StatusCode() == fasthttp.StatusOK {
			ctx.Response.Header.SetStatusCode(fasthttp.StatusInternalServerError)
		}

		if tx, ok := ctx.UserValue(txKey).(*apmfasthttp.TxCloser); ok {
			m.recovery(ctx, tx.Tx(), tx.BodyCapturer(), err)
		}
	}
}
```

Reply:
the recovery function need the transaction and the body capturer.

Reply:
I see. So the need here is to access the Transaction and BodyCapturer. I don't think that's a good enough reason to expose TxCloser, whose job is to close transactions and just happens to expose the transaction and body capturer.

You can already access the transaction with `apm.TransactionFromContext(ctx)`. You now just need a way to get the BodyCapturer. The BodyCapturer is already returned from StartTransactionWithBody, so you can store that in your middleware. I think this need is quite specific to your Atreugo middleware's needs, and not something other users of apmfasthttp will need.

Reply:
I can see the function

```go
func bodyCapturerFromContext(ctx context.Context) *BodyCapturer {
	value, _ := ctx.Value(bodyCapturerKey{}).(*BodyCapturer)
	return value
}
```

That it so similar to:
```go
func TransactionFromContext(ctx context.Context) *Transaction {
	value, _ := apmcontext.TransactionFromContext(ctx).(*Transaction)
	return value
}
```

but the `bodyCapturerFromContext ()` is private and can't be overwrite like `TransactionFromContext ()`, if the `bodyCapturerFromContext ()` were public and overwritable, i can save it into the `fasthttp.RequestCtx` and get  the BodyCapturer with `apm.BodyCapturerFromContext`like the transaction with `apm.TransactionFromContext`.

Also, when you call to `apm.CaptureError ()`, internally call to `bodyCapturerFromContext ()` that now always return `nil` because the apmfasthttp not save the bodycapturer with` ContextWithBodyCapturer () `

Reply:
If i save the returned BodyCapturer of `apmfasthttp.StartTransactionWithBody()` in the framework middleware, each framework implementation must save that.

If `fasthttp.RequestCtx` support `context.Context` should be easier, but it's not like that.

Reply:
@axw Please check this commit: https://github.com/elastic/apm-agent-go/pull/957/commits/dd6b9a1c4237a71fac31cb99439b5616b6a60ddf. if you like this solution.

Reply:
Now, i can get the body capture easily:

```go
func (m *Middleware) PanicView() atreugo.PanicView {
	return func(ctx *atreugo.RequestCtx, err interface{}) {

		if m.panicPropagation {
			defer panic(err)
		}

		// 500 status code will be set only for APM transaction
		// to allow other middleware to choose a different response code
		if ctx.Response.Header.StatusCode() == fasthttp.StatusOK {
			ctx.Response.Header.SetStatusCode(fasthttp.StatusInternalServerError)
		}

		tx := apm.TransactionFromContext(ctx)
		bc := apm.BodyCapturerFromContext(ctx)

		m.recovery(ctx, tx, bc, err)
	}
}
```

Reply:
This looks perfect, thank you! :smile:

Reply:
Nice! 🎉